### ISSUE TITLE
Feature Request: equivalent of `kubectl patch`

### ISSUE THREAD

**Issue by zapman449** [terraform-providers/terraform-provider-kubernetes#723]
### Terraform Version
Terraform v0.12.18

### Affected Resource(s)
n/a (request for new resource)

In AWS EKS, clusters come "pre-configured" with several things running in the `kube-system` namespace.  We need to patch those pre-configured things, while retaining any "upstream" changes which happen to be made.  (for example: set HTTP_PROXY variables)

kubectl provides the `patch` keyword to handle this use-case.  

The kubernetes provider for terraform should do the same.

Proposed example (this would add the `proxy-environment-variables` ConfigMap to the existing `envFrom` list which already contains `aws-node-environment-variable-additions` for the container named `aws-node`):

```
resource "kubernetes_patch" "aws-node" {
  kind = daemonset
  metadata {
    name      = "aws-node"
    namespace = "kube-system"
  }
  spec {
    template {
      spec {
        container {
          name = "aws-node"
          envFrom {
            [
              configMapRef {
                name: proxy-environment-variables
              }
              configMapRef {
                name: aws-node-environment-variable-additions
              }
            ]
          }
        }
      }
    }
  }
}
```

**Comment by ronaldmiranda** [comment]
+1

**Comment by antonosmond** [comment]
I have 2 additional use cases for the same feature, both on EKS.

1. If you want to utilise node taints & tolerations for all your nodes, any EKS managed k8s resources e.g. coredns must be patched to tolerate the taints.

2. Fargate on EKS. If you want to run a nodeless cluster and use Fargate to run everything, some EKS managed resources e.g. coredns prevent this via an annotation e.g.
```
  annotations:
    eks.amazonaws.com/compute-type: ec2
```
The annotation must be removed.
The ability to patch resources would solve both these use cases and many others.

**Comment by alexandrsemak** [comment]
> I have 2 additional use cases for the same feature, both on EKS.
> 
> 1. If you want to utilise node taints & tolerations for all your nodes, any EKS managed k8s resources e.g. coredns must be patched to tolerate the taints.
> 2. Fargate on EKS. If you want to run a nodeless cluster and use Fargate to run everything, some EKS managed resources e.g. coredns prevent this via an annotation e.g.
> 
> ```
>   annotations:
>     eks.amazonaws.com/compute-type: ec2
> ```
> 
> The annotation must be removed.
> The ability to patch resources would solve both these use cases and many others.

Yeah also it would be nice to have equivalent of `kubectl taint node` as it will not work w/o taint node before

**Comment by stoimendhristov** [comment]
+1

**Comment by stoimendhristov** [comment]
I am currently trying to update an existing ConfigMap and simply add more rules to it but once the CM created it seems that it cannot be referred to in order to be updated.

Any thoughts?

Thanks

**Comment by adilc** [comment]
+1

**Comment by Sharathmk99** [comment]
+1

**Comment by eugene-burachevskiy** [comment]
When we setup EKS cluster with terraform and are using tainted on-demand nodes for all system services, we have to patch CoreDNS first to make all further installed apps working. For now we can't patch existing EKS CoreDNS with terraform so we have to install 3rd party CoreDNS helm chart at the beginning.

Ability to patch existing deployments would be really great.

**Comment by hijakebye** [comment]
+1 Would love this for some of our enterprise EKS Fargate Deployments

**Comment by ivan-sukhomlyn** [comment]
It would nice to have such a Terraform resource to patch EKS `aws-node` DaemonSet with a custom ServiceAccount. For example, in the case of the IRSA approach usage for Pods authorization.

**Comment by vide** [comment]
This is also needed to patch EKS clusters hit by https://github.com/kubernetes/kubernetes/issues/61486

**Comment by blawlor** [comment]
This feature would feed very well into things like custom CNI on EKS

**Comment by memory** [comment]
This would also help for management of service meshes such as linkerd or istio, where one might want to add annotations to control mesh proxy injection into the `kube-system` or `default` namespace.

This request is actually being made in different forms in several issues now, see also:

https://github.com/hashicorp/terraform-provider-kubernetes/issues/238
https://github.com/hashicorp/terraform/issues/22754

**Comment by memory** [comment]
For anyone else who's running into this, we've for the moment worked around it with a truly awful abuse of the null resource and local provisioner:

```
resource "null_resource" "k8s_patcher" {
  triggers = {
    // fire any time the cluster is update in a way that changes its endpoint or auth
    endpoint = google_container_cluster.default.endpoint
    ca_crt   = google_container_cluster.default.master_auth[0].cluster_ca_certificate
    token    = data.google_client_config.provider.access_token
  }

  # download kubectl and patch the default namespace
  provisioner "local-exec" {
    command = <<EOH
cat >/tmp/ca.crt <<EOF
${base64decode(google_container_cluster.default.master_auth[0].cluster_ca_certificate)}
EOF
curl -LO https://storage.googleapis.com/kubernetes-release/release/$(curl -s https://storage.googleapis.com/kubernetes-release/release/stable.txt)/bin/linux/amd64/kubectl && chmod +x ./kubectl
./kubectl \
  --server="https://${google_container_cluster.default.endpoint}" \
  --token="${data.google_client_config.provider.access_token}" \
  --certificate_authority=/tmp/ca.crt \
  patch namespace default \
  -p '{"apiVersion":"v1","kind":"Namespace","metadata":{"annotations":{"linkerd.io/inject":"enabled"},"name":"default"}}'
EOH
  }
}
```

**Comment by Haptr3c** [comment]
I tweaked @memory's null_resource workaround to work with the aws provider. This should save anyone looking to run fargate-only EKS a bit of time.

`resource "aws_eks_fargate_profile" "coredns" {
  cluster_name           = aws_eks_cluster.main.name
  fargate_profile_name   = "coredns"
  pod_execution_role_arn = aws_iam_role.fargate_pod_execution_role.arn
  subnet_ids             = var.private_subnets.*.id
  selector {
    namespace = "kube-system"
    labels = {
      k8s-app = "kube-dns"
    }
  }
}

resource "null_resource" "k8s_patcher" {
  depends_on = [ aws_eks_fargate_profile.coredns ]
  triggers = {
    // fire any time the cluster is update in a way that changes its endpoint or auth
    endpoint = aws_eks_cluster.main.endpoint
    ca_crt   = base64decode(data.aws_eks_cluster.cluster.certificate_authority[0].data)
    token    = data.aws_eks_cluster_auth.cluster.token
  }
  provisioner "local-exec" {
    command = <<EOH
cat >/tmp/ca.crt <<EOF
${base64decode(data.aws_eks_cluster.cluster.certificate_authority[0].data)}
EOF
apk --no-cache add curl && \
curl -o aws-iam-authenticator https://amazon-eks.s3.us-west-2.amazonaws.com/1.17.9/2020-08-04/bin/linux/amd64/aws-iam-authenticator && chmod +x ./aws-iam-authenticator && \
curl -LO https://storage.googleapis.com/kubernetes-release/release/$(curl -s https://storage.googleapis.com/kubernetes-release/release/stable.txt)/bin/linux/amd64/kubectl && chmod +x ./kubectl && \
mkdir -p $HOME/bin && mv ./aws-iam-authenticator $HOME/bin/ && export PATH=$PATH:$HOME/bin && \
./kubectl \
  --server="${aws_eks_cluster.main.endpoint}" \
  --certificate_authority=/tmp/ca.crt \
  --token="${data.aws_eks_cluster_auth.cluster.token}" \
  patch deployment coredns \
  -n kube-system --type json \
  -p='[{"op": "remove", "path": "/spec/template/metadata/annotations/eks.amazonaws.com~1compute-type"}]'
EOH
  }
}`

**Comment by Z3R6** [comment]
Any updates?

**Comment by kvaidas** [comment]
The following might also be a viable workaround:
```
resource "local_file" "kubeconfig" {
  filename = pathexpand("~/.kube/config")
  content = <<-CONFIG
    apiVersion: v1
    kind: Config
    clusters:
    - name: clustername
      cluster:
        server: ${aws_eks_cluster.this.endpoint}
        certificate-authority-data: ${aws_eks_cluster.this.certificate_authority.0.data}
    contexts:
    - name: contextname
      context:
        cluster: clustername
        user: username
    current-context: contextname
    users:
    - name: username
      user:
        token: ${data.aws_eks_cluster_auth.this-auth.token}
  CONFIG
}
```

Might work quicker since the token should only be requested once and then reused for any kubectl commands.

Also doesn't depend on having aws-cli installed.

**Comment by vdahmane** [comment]
Any update?

**Comment by johanferguth** [comment]
Same question for me :) : 
If I want to allow other users than me to manage an Aws Eks cluster, I have to edit the configmap aws-auth. It could be very useful to patch this configmap after a deployment rather than replace it totally.

**Comment by ams0** [comment]
Adding to the list, patching `argocd-cm` ConfigMap to add a private repository. I bootstrap AKS+ArgoCD and I'd like to use a private repos for the apps.

**Comment by GolubevV** [comment]
Same issue here, with the need to patch the coredns for taints-tollerations setup and aws-node daemonset for some parameters tweaking (like IP warm target and external SNAT enable).

Will be really nice to be able to get all resources provisioned in one shot by terraform without workarrounds like `local-exec` provisioner which does not work on TFE out of the box due to missing kubectl.

**Comment by holgerson97** [comment]
This is also relevant when you want to deploy an EKS cluster only running Fargate. you need to patch the existing CoreDNS deployment in order to deploy it as Fargate.

**Comment by shanoor** [comment]
Also needed to simply edit the coredns-custom configmap that is created by default in AKS.

**Comment by adamrushuk** [comment]
> Adding to the list, patching `argocd-cm` ConfigMap to add a private repository. I bootstrap AKS+ArgoCD and I'd like to use a private repos for the apps.

I've got a similar requirement, so until there is a better method, I'm using a template and null resource:
```bash
# argocd-cm patch
# https://registry.terraform.io/providers/hashicorp/template/latest/docs/data-sources/file
data "template_file" "argocd_cm" {
  template = file(var.argocd_cm_yaml_path)
  vars = {
    tenantId    = data.azurerm_client_config.current.tenant_id
    appClientId = azuread_service_principal.argocd.application_id
  }
}

# https://www.terraform.io/docs/provisioners/local-exec.html
resource "null_resource" "argocd_cm" {
  triggers = {
    yaml_contents = filemd5(var.argocd_cm_yaml_path)
    sp_app_id     = azuread_service_principal.argocd.application_id
  }

  provisioner "local-exec" {
    interpreter = ["/bin/bash", "-c"]
    environment = {
      KUBECONFIG = var.aks_config_path
    }
    command = <<EOT
      kubectl patch configmap/argocd-cm --namespace argocd --type merge --patch "${data.template_file.argocd_cm.rendered}"
    EOT
  }

  depends_on = [
    local_file.kubeconfig,
    null_resource.argocd_configure
  ]
}
```

**Comment by eahrend** [comment]
Adding a use case to this list as well, I'm on GKE and I'm using the terraform code below to create a namespace/secrets to help bootstrap a cluster with service account keys required by the applications. If the secret data changes, I'd like to overwrite the secret, but this fails due to the namespace existing already.

```hcl
resource "kubernetes_namespace" "default" {
  metadata {
    name = var.namespace
  }
}


resource "kubernetes_secret" "default" {
  depends_on = [kubernetes_namespace.default]
  metadata {
    name = var.kube_secret_name
    namespace = var.namespace
  }
  data = var.secret_data
  type = var.secret_type
}
```

**Comment by jrhouston** [comment]
Thanks everyone for your patience on this issue.

We're looking at implementing this feature and are discussing where it should live and what its implementation is going to look like. If you want to add your 2¢ please feel free to contribute your thoughts to this proposal PR: 

https://github.com/hashicorp/terraform-provider-kubernetes/pull/1257

**Comment by bfelaco** [comment]
+1 on this feature request for a generic way to apply patches to existing resources. It would also be helpful to have a prescriptive way to handle the common requirement to create/update specific key/values within an existing `ConfigMap` or `Secret`.

**Comment by alexmnyc** [comment]
Please note ability to add annotation to existing resources like service account, there should be 

resource "kubernetes_service_account" {
    allow_import_if_exists = true
    metadata {
       name = "test_sa"
       annotation  = "test_annotation"
   }
}

**Comment by jrhouston** [comment]
Thanks for adding your thoughts @bfelaco @alexmnyc.

I'm collecting acceptance test criteria for this feature. If you have a specific use case you need this for please share with as much detail as possible and add here to the proposal PR linked above. If you are already solving it outside or Terraform or with a `null_resource` I'd love to hear about that too.

**Comment by herrLierb** [comment]
[deamon-set-patch.zip](https://github.com/hashicorp/terraform-provider-kubernetes/files/6517594/deamon-set-patch.zip)

I can share a sample using null-resources to enable IRSA for the node daemonset as described here: https://aws.github.io/aws-eks-best-practices/security/docs/iam/#update-the-aws-node-daemonset-to-use-irsa

many thanks to @jensbac ! :)

**Comment by mbevc1** [comment]
Adding myself to notification :+1: this one

**Comment by RichiCoder1** [comment]
> Adding myself to notification 👍 this one

For future reference, you can just "Subscribe" to the issue.

**Comment by Sebor** [comment]
One more use case of AWS EKS: there is a default storage class "gp2" when I create cluster. But during cluster provisioning I create another storage class ("gp3", for example) and I want to set it as default. To do that firstly I have to set annotation "storageclass.kubernetes.io/is-default-class: `false`" for "gp2" and only after that create new storage class with annotation "storageclass.kubernetes.io/is-default-class: `true`".

Currently I do that via local-provisioner:

```hcl
  provisioner "local-exec" {
    command = "kubectl patch storageclass gp2 -p '{\"metadata\": {\"annotations\":{\"storageclass.kubernetes.io/is-default-class\":\"false\"}}}'"

    environment = {
      KUBECONFIG = module.eks_cluster.kubeconfig_filename
    }
  }
```

It's ugly and I must have local kubeconfig file

**Comment by scalen** [comment]
Flagging the details of another usecase in AWS EKS: enabling CNI support for Pod-level Security Groups/secure networking.  [This requires setting an environment variable in the template of the `aws-node` DaemonSet](https://www.eksworkshop.com/beginner/115_sg-per-pod/30_cni_config/), in the `kube-system` namespace.

Ideally, this would remain managed EKS in general, and just have that environment variable changed by Terraform.

**Comment by aleixrm-scopely** [comment]
Another use case, similar to @scalen 's one, is patching the aws-cni-plugin add-on for changing the WARM_ENI_TARGET parameter setting an environment variable  (https://github.com/aws/amazon-vpc-cni-k8s/blob/master/docs/eni-and-ip-target.md)

**Comment by artificial-aidan** [comment]
Added a bit of configurability to my patch resource

```
data "aws_eks_cluster" "cluster" {
  name = var.cluster_name
}

data "aws_eks_cluster_auth" "cluster" {
  name = var.cluster_name
}

resource "local_file" "config" {
  content  = ""
  filename = "${path.module}/.terraform/config.yaml"
}

resource "local_file" "kube_ca" {
  content  = base64decode(data.aws_eks_cluster.cluster.certificate_authority[0].data)
  filename = "${path.module}/.terraform/ca.crt"
}

resource "null_resource" "patch_cni" {
  triggers = {
    always_run = "${timestamp()}"
  }

  provisioner "local-exec" {
    command = "kubectl set env ds aws-node --server=$KUBESERVER --token=$KUBETOKEN --certificate-authority=$KUBECA -n kube-system ENABLE_PREFIX_DELEGATION=true WARM_PREFIX_TARGET=1"

    environment = {
      KUBECONFIG = local_file.config.filename
      KUBESERVER = data.aws_eks_cluster.cluster.endpoint
      KUBETOKEN  = data.aws_eks_cluster_auth.cluster.token
      KUBECA     = local_file.kube_ca.filename
    }
  }
}

```

**Comment by scalen** [comment]
> A equivalent of kubectl patch could be a partial apply with the kubernetes_manifest resource that only updates the fields defined inside the resource. In combination with the kubernetes service-side apply feature, only the fields should be update managed by the field-manager terraform.

This is definitely true, in some contexts, but it seems to depend on how the Server-Side Apply features are invoked.  For example, it isn't clear (to me) from the documentation if incomplete resources are supported via all channels.  Does the current behaviour of `kubernetes_manifest` (and/or the other `kubernetes_*` resources) support such incomplete manifests? Or is there an argument available to set to support partial applies?  Either way, clarification of these in the documentation would be very useful!

**Comment by jkroepke** [comment]
A second approach would be a new resource like kubernetes_manifest_field that manages only one field of an existing resource. That could be the more terraform way.

**Comment by Maxwell2022** [comment]
With the `null_resource` work around, I found that using kubeconfig template looks a bit more clean ([source](https://github.com/GSA/terraform-kubernetes-aws-load-balancer-controller/blob/main/main.tf#L428-L499))

```hcl
data "template_file" "kubeconfig" {
  template = <<-EOF
    apiVersion: v1
    kind: Config
    current-context: terraform
    clusters:
    - name: ${aws_eks_cluster.cluster.name}
      cluster:
        certificate-authority-data: ${aws_eks_cluster.cluster.certificate_authority.0.data}
        server: ${aws_eks_cluster.cluster.endpoint}
    contexts:
    - name: terraform
      context:
        cluster: ${aws_eks_cluster.cluster.name}
        user: terraform
    users:
    - name: terraform
      user:
        token: ${data.aws_eks_cluster_auth.cluster.token}
  EOF
}

resource "null_resource" "update_coredns_annotations" {
  triggers = {
    kubeconfig = base64encode(data.template_file.kubeconfig.rendered)
    cmd_patch = <<-EOF
      cat <<YAML | kubectl \
        -n kube-system \
        --kubeconfig <(echo $KUBECONFIG | base64 --decode) \
        patch deployment coredns \
        --type json \
        -p='[{"op": "remove", "path": "/spec/template/metadata/annotations/eks.amazonaws.com~1compute-type"}]'
    EOF
  }

  provisioner "local-exec" {
    interpreter = ["/bin/bash", "-c"]
    environment = {
      KUBECONFIG = self.triggers.kubeconfig
    }
    command = self.triggers.cmd_patch
  }

  depends_on = [aws_eks_fargate_profile.coredns, aws_eks_cluster.cluster]
}
```

**Comment by jkroepke** [comment]
interesting, but not cross os compatible.

**Comment by robertb724** [comment]
Here is the workaround we came up with, this will run as soon as the cluster is created in our case:

```

resource "kubernetes_job" "kube-system-labeler" {
  metadata {
    name      = "kube-system-labeler"
    namespace = utility
  }
  spec {
    template {
      metadata {}
      spec {
        service_account_name = utility
        container {
          name    = "ns-labeler"
          image   = "google/cloud-sdk:latest"
          command = ["/bin/sh", "-c", "kubectl label ns kube-system --overwrite config.linkerd.io/admission-webhooks=disabled admission.gatekeeper.sh/ignore=no-self-managing"]
        }
        restart_policy = "Never"
      }
    }
  }
  wait_for_completion = true
  timeouts {
    create = "5m"
  }
}
```

**Comment by watsonjm** [comment]
This is an issue for us as well since we frequently do work in AWS EKS where other users need to be added to aws-auth configmap, but this is not currently possible without external dependencies (kubectl).

**Comment by shinebayar-g** [comment]
@watsonjm fyi you can manually add `aws-auth` configmap before it's created automatically for you. You have to create this configmap in terraform before you create your first nodepool. https://docs.aws.amazon.com/eks/latest/userguide/add-user-role.html

**Comment by marcuz** [comment]
> This is an issue for us as well since we frequently do work in AWS EKS where other users need to be added to aws-auth configmap, but this is not currently possible without external dependencies (kubectl).

On top of this, since release 18 of [terraform-aws-modules/terraform-aws-eks](https://github.com/terraform-aws-modules/terraform-aws-eks), `aws-auth` isn't managed by the module anymore, most of the workarounds are based on exec/kubectl which is not something everyone can do.

**Comment by aidan-melen** [comment]
> Here is the workaround we came up with, this will run as soon as the cluster is created in our case:
> 
> ```
> 
> resource "kubernetes_job" "kube_system_labeler" {
>   metadata {
>     name      = "kube-system-labeler"
>     namespace = utility
>   }
>   spec {
>     template {
>       metadata {}
>       spec {
>         service_account_name = utility
>         container {
>           name    = "ns-labeler"
>           image   = "bitnami/kubectl:latest"
>           command = ["/bin/sh", "-c", "kubectl label ns kube-system --overwrite config.linkerd.io/admission-webhooks=disabled admission.gatekeeper.sh/ignore=no-self-managing"]
>         }
>         restart_policy = "Never"
>       }
>     }
>   }
>   wait_for_completion = true
>   timeouts {
>     create = "5m"
>   }
> }
> ```

Can you please describe how this is a solution?

**Comment by robertb724** [comment]
What we are trying to solve is managing the resources managed by the cloud provider. In my case, GKE itself manages the `kube-system`  namespace.  We are installing `linkerd`, which requires that the `kube-system` namespace has the annotation `config.linkerd.io/admission-webhooks=disabled`. We can not manage the `kube-system` as a tf resource since it would collide with what gke has created. This spins up a pod as soon as the cluster is ready(since it is using the k8s provider backed by this cluster) to annotate the namespace for us. It will fail if the annotation is not there and we can control when it is executed, just as if we were able to interact with the namespace directly from the terraform execution.

**Comment by fabienpelletier** [comment]
I'm also hitting this issue when deploying `tigera-operator` on a new cluster.
Enabling the Wireguard encryption[0] requires me to do 
```kubectl patch felixconfiguration default --type='merge' -p '{"spec":{"wireguardEnabled":true}}'```
which I am currently running in a `null_resource`'s `local-provisioner`.
 
 [0] https://projectcalico.docs.tigera.io/security/encrypt-cluster-pod-traffic

**Comment by aidanmelen** [comment]
> A second approach would be a new resource like `kubernetes_manifest_field` that manages only one field of an existing resource. That could be the more terraform way.

This would be a nice "terraformic" solution for an ugly problem caused by terraform needing to partially manage objects.

**Comment by aidanmelen** [comment]
@zapman449 @memory @vide @blawlor @aareet upvote [my PR](https://github.com/hashicorp/terraform-provider-kubernetes/pull/1667) ^^^ so that we can get some traction on this. I got patch functionality working for the `kubernetes_manifest` resource.

**Comment by jrhouston** [comment]
For those of you whose use-case is patching labels, annotations, and ConfigMap entries v2.10.0 of the provider brought support for doing this using Server-Side Apply & Field Manager in the following resources:

- [kubernetes_labels](https://registry.terraform.io/providers/hashicorp/kubernetes/latest/docs/resources/labels)
- [kubernetes_annotations](https://registry.terraform.io/providers/hashicorp/kubernetes/latest/docs/resources/annotations)
- [kubernetes_config_map_v1_data](https://registry.terraform.io/providers/hashicorp/kubernetes/latest/docs/resources/config_map_v1_data)

Other use-cases on our radar for resources where Terraform will partially manage a Kubernetes resource:
- Adding container environment variables
- Setting taints and tolerations

If you have another use-use please share it. 

For some context on why we haven't added a completely generic patch resource see [this discussion](https://github.com/hashicorp/terraform-provider-kubernetes/pull/1257).

**Comment by bryantbiggs** [comment]
@jrhouston it would be great if we could update/remove an annotation if possible. In order to run a Fargate only EKS cluster, you have to remove an annotation from the deployment template spec using: 

```bash
kubectl patch deployment coredns \
    -n kube-system \
    --type json \
    -p='[{"op": "remove", "path": "/spec/template/metadata/annotations/eks.amazonaws.com~1compute-type"}]'
```

See more here https://docs.aws.amazon.com/eks/latest/userguide/fargate-getting-started.html#fargate-gs-coredns and there is a small convo started https://github.com/aws-samples/aws-eks-accelerator-for-terraform/issues/394

**Comment by jkroepke** [comment]
Did you try https://registry.terraform.io/providers/hashicorp/kubernetes/latest/docs/resources/annotations and set the value to null? It should remove the annotation.

**Comment by bryantbiggs** [comment]
> Did you try [registry.terraform.io/providers/hashicorp/kubernetes/latest/docs/resources/annotations](https://registry.terraform.io/providers/hashicorp/kubernetes/latest/docs/resources/annotations) and set the value to null? It should remove the annotation.

@jkroepke yes, that was the resource I was using but I couldn't figure out how to get the resource to target the `spec.template.metadata` annotations. 

Using it like below just lets me access the deployments annotations, but I need to modify the the pod specs annotations:

```hcl
resource "kubernetes_annotations" "coredns" {
  api_version = "apps/v1"
  kind        = "Deployment"
  
  metadata {
    name      = "coredns"
    namespace = "kubesystem"
  }

  annotations = {}
}
```

**Comment by tanvp112** [comment]
According to the `annotations` document:

> This resource uses [field management](https://kubernetes.io/docs/reference/using-api/server-side-apply/#field-management) and [server-side apply](https://kubernetes.io/docs/reference/using-api/server-side-apply/) to manage **only the annotations that are defined in the Terraform configuration**.

The annotation is not managed in Terraform configuration, not sure if the module is applicable to EKS **managed** addon like coreDNS.

**Comment by jrhouston** [comment]
> Using it like below just lets me access the deployments annotations, but I need to modify the the pod specs annotations

@bryantbiggs Yeah this isn't going to work because the resource is an analog to `kubectl annotate` – it doesn't patch deeper nested metadata just the top level. I think that's something we're going to need to add a separate resource for, thanks for surfacing this use-case.

**Comment by aidanmelen** [comment]
From my work in this PR

- https://github.com/hashicorp/terraform-provider-kubernetes/pull/1667

I found that the `kuberenetes_manifest` resource was erroring out on patching already existing resources because simple condition in the `apply.go` code. Commenting out [this if statement](https://github.com/hashicorp/terraform-provider-kubernetes/blob/main/manifest/provider/apply.go#L303-L322) would add support for patching already existing resources. I was pleasantly surprised that it worked for the tests that I ran. I believe this worked because the `kuberentes_manifest` resource already refreshes the k8s resource during the first plan which essentially acts as an import.

There are plenty of examples where bringing already existing k8s resource under control of this terraform provider would be extremely useful, `aws-auth`, `coredns`, etc. So I propose we add a new experimental `kuberenetes_patch` resource, as the OP suggested. This code would be trivial to maintain as it would be reusing`kuberneters_manifest` resource code... except for the already exists condition.

Normally I would protest breaking the fundamentals of terraform CRUD, but since there is an analog in `kubectl patch`; it will be intuitive for users to understand how this special resource would behave. Since it will be a dedicated resource we can thoroughly document the special lifecycle behavior to avoid any confusion.

Thoughts? Happy to make another PR if that would help?

**Comment by jeroenj** [comment]
> If you have another use-case please share it.

Our use-case is actually applying an affinity rule to the `kube-dns` Deployment:

```sh
kubectl patch deployment kube-dns --namespace kube-system --patch "$(cat <<EOF
apiVersion: apps/v1
kind: Deployment
metadata:
  name: kube-dns
  namespace: kube-system
spec:
  template:
    spec:
      affinity:
        nodeAffinity:
          preferredDuringSchedulingIgnoredDuringExecution:
          - weight: 1
            preference:
              matchExpressions:
              - key: workload
                operator: In
                values:
                - stable
EOF
)"
```

**Comment by cmanfre4** [comment]
Here's a more fleshed out workaround that I came up with based on what [robertb724 had suggested](https://github.com/hashicorp/terraform-provider-kubernetes/issues/723#issuecomment-999865153):

```
resource "kubernetes_service_account" "core_dns_fixer" {
  metadata {
    name      = "core-dns-fixer"
    namespace = "kube-system"
  }
}

resource "kubernetes_role" "core_dns_fixer" {
  metadata {
    name = "core-dns-fixer"
    namespace = "kube-system"
  }

  rule {
    api_groups     = ["apps"]
    resources      = ["deployments"]
    resource_names = ["coredns"]
    verbs          = ["get", "patch"]
  }
}

resource "kubernetes_role_binding" "core_dns_fixer" {
  metadata {
    name      = "core-dns-fixer"
    namespace = "kube-system"
  }
  role_ref {
    api_group = "rbac.authorization.k8s.io"
    kind      = "Role"
    name      = kubernetes_role.core_dns_fixer.metadata[0].name
  }
  subject {
    kind      = "ServiceAccount"
    name      = kubernetes_service_account.core_dns_fixer.metadata[0].name
    namespace = "kube-system"
  }
}

resource "kubernetes_job" "patch_core_dns" {
  depends_on = [
    # assumes the `kube-system` fargate profile is created in the same code
    # edit to match yours, or comment out if it's being created elsewhere
    aws_eks_fargate_profile.main["kube-system"],
    kubernetes_role_binding.core_dns_fixer
  ]
  metadata {
    name      = "patch-core-dns"
    namespace = "kube-system"
  }
  spec {
    template {
      metadata {}
      spec {
        service_account_name = kubernetes_service_account.core_dns_fixer.metadata[0].name
        container {
          name    = "patch-core-dns"
          image   = "bitnami/kubectl:latest"
          command = ["/bin/sh", "-c", "kubectl patch deployments.app/coredns -n kube-system --type json -p='[{\"op\": \"remove\", \"path\": \"/spec/template/metadata/annotations/eks.amazonaws.com~1compute-type\"}]'"]
        }
        restart_policy = "Never"
      }
    }
  }
  wait_for_completion = true
  timeouts {
    create = "5m"
  }
}

resource "kubernetes_job" "restart_core_dns" {
  depends_on = [
    kubernetes_job.patch_core_dns
  ]
  metadata {
    name      = "restart-core-dns"
    namespace = "kube-system"
  }
  spec {
    template {
      metadata {}
      spec {
        service_account_name = kubernetes_service_account.core_dns_fixer.metadata[0].name
        container {
          name    = "restart-core-dns"
          image   = "bitnami/kubectl:latest"
          command = ["/bin/sh", "-c", "kubectl rollout restart deployments.app/coredns -n kube-system"]
        }
        restart_policy = "Never"
      }
    }
  }
  wait_for_completion = true
  timeouts {
    create = "5m"
  }
}
```

This will create a service account that is bound to a role that has permissions to patch and restart the coredns deployment in the kube-system namespace. It will create two one-time Kubernetes jobs that use the aforementioned service account; one to perform the patch, and then one to restart the deployment.

If you want to run this along with the same terraform code as your eks cluster is being created in, you can configure the Kubernetes provider with something like the following:

```
data "aws_eks_cluster" "main" {
  name = aws_eks_cluster.main.name
}

data "aws_eks_cluster_auth" "main" {
  name = aws_eks_cluster.main.name
}

provider "kubernetes" {
  host                   = data.aws_eks_cluster.main.endpoint
  cluster_ca_certificate = base64decode(data.aws_eks_cluster.main.certificate_authority[0].data)
  token                  = data.aws_eks_cluster_auth.main.token
}
```

Note: it assumes that you've named your `aws_eks_cluster` resource as `main`. Also refer to the comments in the explicit `depends_on` in the `kubernetes_job.patch_core_dns` resource. It needs to have the fargate profile for the `kube-system` namespace to be created or it will not be able to run.

**Comment by FernandoMiguel** [comment]
That's very creative @cmanfre4

**Comment by jamesanto** [comment]
Expanding on @cmanfre4's answer, we could probably simplify it to a single job with this command:

```["/bin/sh", "-c", "compute_type=$(kubectl get deployments.app/coredns -n kube-system -o jsonpath='{.spec.template.metadata.annotations.eks\\.amazonaws\\.com/compute-type}'); [ ! -z \"$compute_type\" ] && kubectl patch deployments.app/coredns -n kube-system --type json -p='[{\"op\":\"remove\", \"path\": \"/spec/template/metadata/annotations/eks.amazonaws.com~1compute-type\"}]' && kubectl rollout restart deployments.app/coredns -n kube-system"]```

**Comment by cmanfre4** [comment]
The reason I split it in to two discrete terraform resources is because if the patch succeeds but the restart fails, terraform can retry the restart by recreating the tainted `kubernetes_job.restart_core_dns` resource on a subsequent apply. If both commands are in one job, the patch will fail if tried a second time.

**Comment by bryantbiggs** [comment]
Just FYI - if you patch the CoreDNS on EKS, you'll want to eject from the EKS API managing the CoreDNS deployment using the `preserve = true`. If not, the next time the EKS API updates the addon, it will remove your patch and cause your DNS to fail

**Comment by FernandoMiguel** [comment]
> Thanks @jkroepke!
> 
> I was able to remove the default storage class from an EKS cluster with https://registry.terraform.io/providers/hashicorp/kubernetes/latest/docs/resources/annotations.
> 
> ```terraform
> resource "kubernetes_annotations" "default-storageclass" {
>   api_version = "storage.k8s.io/v1"
>   kind        = "StorageClass"
>   force       = "true"
> 
>   metadata {
>     name = "gp2"
>   }
>   annotations = {
>     "storageclass.kubernetes.io/is-default-class" = "false"
>   }
> }
> ```

that's very interesting way of doing this

**Comment by dudicoco** [comment]
I have a very simple workaround in the form of a bootstrap script which deletes the relevant default resources:
```sh
#!/usr/bin/env bash
# shellcheck disable=SC2043

set -euo pipefail

while test $# -gt 0; do
  case "$1" in
  -h | --help)
    echo " "
    echo "options:"
    echo "-h, --help            show brief help"
    echo "--context             specify kube contxt"
    exit 0
    ;;
  --context)
    shift
    if test $# -gt 0; then
      context=$1
    else
      echo "no kube context specified"
      exit 1
    fi
    shift
    ;;
  *)
    break
    ;;
  esac
done

for kind in daemonset clusterRole clusterRoleBinding serviceAccount; do
  echo "deleting $kind/aws-node"
  kubectl --context "$context" --namespace kube-system delete $kind aws-node
done

for kind in customResourceDefinition; do
  echo "deleting $kind/eniconfigs.crd.k8s.amazonaws.com"
  kubectl --context "$context" --namespace kube-system delete $kind eniconfigs.crd.k8s.amazonaws.com
done

for kind in daemonset serviceAccount; do
  echo "deleting $kind/kube-proxy"
  kubectl --context "$context" --namespace kube-system delete $kind kube-proxy
done

for kind in configMap; do
  echo "deleting $kind/kube-proxy-config"
  kubectl --context "$context" --namespace kube-system delete $kind kube-proxy-config
done

for kind in deployment serviceAccount configMap; do
  echo "deleting $kind/coredns"
  kubectl --context "$context" --namespace kube-system delete $kind coredns
done

for kind in service; do
  echo "deleting $kind/kube-dns"
  kubectl --context "$context" --namespace kube-system delete $kind kube-dns
done

for kind in storageclass; do
  echo "deleting $kind/gp2"
  kubectl --context "$context" delete $kind gp2
done
```

**Comment by Adiii717** [comment]
I've got a similar requirement to update the ArgoCD password, and this worked for me

```hcl
resource "null_resource" "argocd_update_pass" {
  provisioner "local-exec" {
    interpreter = ["/bin/bash", "-c"]
    command     = <<EOT
    kubectl patch secret -n argocd argocd-secret -p '{"stringData": { "admin.password": "'$(htpasswd -bnBC 10 "" ${data.azurerm_key_vault_secret.argocd-password.value} | tr -d ':\n')'"}}'  --kubeconfig ./temp/kube-config.yaml;
    EOT
  }
  depends_on = [
    helm_release.argocd,
    local_file.kube_config
  ]
}

resource "local_file" "kube_config" {
  content  = azurerm_kubernetes_cluster.aks.kube_config_raw
  filename = "${path.module}/temp/kube-config.yaml"
}
```

**Comment by simonvanderveldt** [comment]
> If you have another use-case please share it.

@jrhouston We have another usecase. We're running on GKE with Calico enabled, this gives us a lot of readiness/liveness probe failures because the timeout is set to 1s. This is fixed in Calico (see https://github.com/projectcalico/calico/issues/5122#issuecomment-982865898) but the version of Calico that includes this fix isn't available on (stable) GKE yet. So we want to apply a patch to increase the timeout to match the value as set by newer Calico versions.

Also whilst I understand the desire of Terraform to keep it simple with regards to the implementation, conceptually matching kubectl will probably be simpler for most users to understand and there'd be no need for a dozen or so specific resources.

**Comment by Thibault-Brocheton** [comment]
Here's a use case:
I'm using EKS created by terraform terraform-aws-modules/terraform-aws-eks module
I have different types of self managed node groups in my cluster, some small ec2 called "admins" handling system pods (coredns,  autoscaler, alb constroller, ...) and some large ec2 called "applications" that handle my business applications
I'm looking to automatically update the coredns deployment, created by EKS, so it's nodeSelector target my admins pods

I would love some kind of _kubernetes_node_selector_ to do this patch, instead of having to workaround from a bash command or manually importing coredns after my eks creation

**Comment by b-a-t** [comment]
> > This is an issue for us as well since we frequently do work in AWS EKS where other users need to be added to aws-auth configmap, but this is not currently possible without external dependencies (kubectl).
> 
> On top of this, since release 18 of [terraform-aws-modules/terraform-aws-eks](https://github.com/terraform-aws-modules/terraform-aws-eks), `aws-auth` isn't managed by the module anymore, most of the workarounds are based on exec/kubectl which is not something everyone can do.

Well, parameters to the module like [manage_aws_auth_configmap](https://github.com/terraform-aws-modules/terraform-aws-eks#input_manage_aws_auth_configmap) suggest otherwise...

**Comment by StephanX** [comment]
Thanks @cmanfre4 for the [tip](https://github.com/hashicorp/terraform-provider-kubernetes/issues/723#issuecomment-1130887972).  I repurposed your solution to replace my default EKS gp2 storage class (which is unencrypted by default.)  I also added a variable `cluster_bootstrap` so that the job only needs to run the first time, while the replacement `gp2` storage class is still managed by terraform:

`terraform apply -var='cluster_bootstrap=true'`

```
resource "kubernetes_service_account" "replace_storage_class_gp2" {
  metadata {
    name      = "replace-storage-class-gp2"
    namespace = "kube-system"
  }
}

resource "kubernetes_cluster_role" "replace_storage_class_gp2" {
  metadata {
    name = "replace-storage-class-gp2"
  }

  rule {
    api_groups     = ["storage.k8s.io" ]
    resources      = ["storageclasses"]
    resource_names = ["gp2"]
    verbs          = ["get", "delete"]
  }
}

resource "kubernetes_cluster_role_binding" "replace_storage_class_gp2" {
  metadata {
    name      = "replace-storage-class-gp2"
  }
  role_ref {
    api_group = "rbac.authorization.k8s.io"
    kind      = "ClusterRole"
    name      = kubernetes_cluster_role.replace_storage_class_gp2.metadata[0].name
  }
  subject {
    kind      = "ServiceAccount"
    name      = kubernetes_service_account.replace_storage_class_gp2.metadata[0].name
    namespace = "kube-system"
  }
}

resource "kubernetes_job" "replace_storage_class_gp2" {
  count = var.cluster_bootstrap ? 1 : 0
  depends_on = [
    kubernetes_cluster_role_binding.replace_storage_class_gp2
  ]
  metadata {
    name      = "replace-storage-class-gp2"
    namespace = "kube-system"
  }
  spec {
    template {
      metadata {}
      spec {
        service_account_name = kubernetes_service_account.replace_storage_class_gp2.metadata[0].name
        container {
          name    = "replace-storage-class-gp2"
          image   = "bitnami/kubectl:latest"
          command = ["/bin/sh", "-c", "kubectl delete storageclass gp2"]
        }
        restart_policy = "Never"
      }
    }
  }
  wait_for_completion = true
  timeouts {
    create = "5m"
  }
}

resource "kubernetes_storage_class" "gp2" {
  metadata {
    name = "gp2"
  }
  storage_provisioner = "kubernetes.io/aws-ebs"
  reclaim_policy      = "Delete"
  parameters = {
    encrypted = "true"
    fsType = "ext4"
    type = "gp2"
  }
  depends_on = [
    kubernetes_job.replace_storage_class_gp2
  ]
}
```

**Comment by ArieLevs** [comment]
@bryantbiggs checkout version 2.15.0 released two days ago
it now contains the [kubernetes_env](https://registry.terraform.io/providers/hashicorp/kubernetes/latest/docs/resources/env) resource which can be used exactly for your use case,
i've successfully tested this updating the AWS CNI plugin (daemonset)

**Comment by djmcgreal** [comment]
Hi. I’d like to patch imagePullSecrets into a ServiceAccount.

**Comment by peteneville** [comment]
Hi, I would like to patch a meshconfig and add ingressGateway.

**Comment by benemon** [comment]
Would very much appreciate the ability to `patch`. When bootstrapping Red Hat OpenShift clusters, there are a large number of Day 1 configuration elements - authn / authz, storage, registry configs - where the workflow revolves around patching existing Cluster Resources into the state required.

**Comment by alicancakil** [comment]
Hello @ArieLevs 
How can I use `kubernetes_env` to do something like this?

```
kubectl patch deployment coredns \
    -n kube-system \
    --type json \
    -p='[{"op": "remove", "path": "/spec/template/metadata/annotations/eks.amazonaws.com~1compute-type"}]'
```

I looked at the docs but couldn't get it working. Would you be able to kindly provide an example?

**Comment by z0rc** [comment]
@alicancakil your specific example can be solved by addon's optional configuration. See https://aws.amazon.com/blogs/containers/amazon-eks-add-ons-advanced-configuration/. https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/eks_addon already supports passing configuration values to addon.

**Comment by ArieLevs** [comment]
@alicancakil the `kubernetes_env` will only add environment values to a supported api resource.
i'm not sure the provider yet supports annotations removals (classic patch) but maybe you can use the [kubernetes_annotations](https://registry.terraform.io/providers/hashicorp/kubernetes/latest/docs/resources/annotations) resource, so maybe something like that can help you
```
resource "kubernetes_annotations" "coredns" {
  api_version = "apps/v1"
  kind        = "Deployment"
  metadata {
    name = "coredns"
    namespace = "kube-system"
  }
  # These annotations will be applied to the Pods created by the Deployment
  template_annotations = {
    "eks.amazonaws.com/compute-type" = ""
  }

  force = true
}
```

this will remove the value of `eks.amazonaws.com/compute-type`

**Comment by bryantbiggs** [comment]
to get create a fully serverless EKS Fargate based cluster, you only need to use the addon configuration like @z0rc mentioned. Here is an example https://github.com/clowdhaus/eks-reference-architecture/blob/f37390db1b38d154979cc1aeb4d72ab53929e847/serverless/eks.tf#L13-L15

**Comment by partcyborg** [comment]
We have another use case

When setting up the GKE identity service for enabling OIDC authentication on the k8s API, the [setup instructions](https://cloud.google.com/kubernetes-engine/docs/how-to/oidc#configuring_on_a_cluster) require you to edit a pre-existing `ClientConfig` resource (a CRD provided by GKE) and fill in a bunch of fields.  There does not appear to be any way to configure this using terraform other than the `null_resource` hack.

**Comment by BBBmau** [comment]
We recently merged and released the ability to patch `initContainers` with #2067, let us know if there are any issues when using the new patch attribute.

**Comment by framctr** [comment]
Another use case would be to be able to patch an existing resource created by an operator. For example, if I deploy rancher managed Prometheus and then I want to change the configuration of the `Prometheus` resource.

**Comment by WyriHaximus** [comment]
To add another use case to the list: Patch a priority class name on deployments/statefulsets

**Comment by gorkemgoknar** [comment]
Another use case, edit configurations of EKS provided kube-proxy via patching configmap

**Comment by jarrettprosser** [comment]
Yet another use case: I'd like to be able to patch the `default` storageclass on an AKS cluster to add tags to the created volumes. That would require adding
```
parameters:
  tags: some-tag=some-value
```
Patching would be preferable to creating new storageclasses, as there are already 7 by default.

**Comment by littlejo** [comment]
I created a resource to patch daemonset in my provider:
https://registry.terraform.io/providers/littlejo/cilium/latest/docs/resources/kubeproxy_free

**Comment by Maxwell2022** [comment]
if you bump into this issue looking to fix **coredns**, you can now configure the addon to use **Fargate** using `computeType`.

```hcl
resource "aws_eks_addon" "coredns_amazon_eks_addon" {
  cluster_name                = var.cluster_name
  addon_name                  = "coredns"
  addon_version               = "v1.11.3-eksbuild.1"
  resolve_conflicts_on_create = "OVERWRITE"
  resolve_conflicts_on_update = "PRESERVE"
  configuration_values = jsonencode({
    computeType = "Fargate"
    # @see https://docs.aws.amazon.com/eks/latest/userguide/coredns-autoscaling.html#:r2a:
    autoScaling = {
      enabled     = true,
      minReplicas = var.coredns_min_replicas,
      maxReplicas = var.coredns_max_replicas,
    }
  })
}
```

**Comment by rosmo** [comment]
For a simple patch, like patching a configmap, I've used Terraform's `import` clause to import the existing resource and then the normal resource to patch in variables (eg. patching kube-dns-autoscaler on GKE - example is for OpenTofu 1.9.0 with the provider for_each support):

```
import {
  for_each = local.clusters_with_kube_dns
  to       = kubernetes_config_map.scale_up_kube_dns[each.key]
  id       = "kube-system/kube-dns-autoscaler"
}

resource "kubernetes_config_map" "scale_up_kube_dns" {
  for_each = local.clusters_with_kube_dns
  provider = kubernetes.by_cluster[each.key]

  metadata {
    name      = "kube-dns-autoscaler"
    namespace = "kube-system"
  }

  data = {
    linear = jsonencode({
      coresPerReplica           = 256
      includeUnschedulableNodes = true
      nodesPerReplica           = 8 # default 16
      preventSinglePointFailure = true
      max                       = 15 # No default value
    })
  }
}
```

**Comment by madchap** [comment]
My use-case is with Cilium, where I would need to patch the VPC CNI with `kubectl -n kube-system patch daemonset aws-node --type='strategic' \
  -p='{"spec":{"template":{"spec":{"nodeSelector":{"io.cilium/aws-node-enabled":"true"}}}}}'` without ugliness upon a fresh EKS cluster creation.

**Comment by moss2k13** [comment]
@madchap I'm just making sure you've seen this:
https://eksctl.io/usage/addons/#cluster-creation-flexibility-for-default-networking-addons
You can install EKS without default network addons so no need to patch/disable aws-node for cilium anymore :wink:

**Comment by madchap** [comment]
> [@madchap](https://github.com/madchap) I'm just making sure you've seen this: https://eksctl.io/usage/addons/#cluster-creation-flexibility-for-default-networking-addons You can install EKS without default network addons so no need to patch/disable aws-node for cilium anymore 😉

ah nice, it slipped by indeed. Will see how I can configure it through https://github.com/terraform-aws-modules/terraform-aws-eks

**Comment by mbevc1** [comment]
Bump
